nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+
+interface Argument {
+  type: ArgumentDataType;
+  value: any;
+}
+
+type ArgumentDataType = 'String'|'Boolean'|'Hash160'|'Hash256'|'Integer'|'ByteArray'|'Array'|'Address';
+</pre>
+
+
+'''Success return value'''
+
+result: Object - Response from RPC node
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`RPC_ERROR`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+invokeRead({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  operation: 'calculatorAdd',
+  arguments: [
+    {
+      type: 'integer',
+      value: 2
+    },
+    {
+      type: 'integer',
+      value: 10
+    }
+  ]
+})
+.then((result: Object) => {
+  console.log('Read invocation result: ' + JSON.stringigy(result));
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_REFUSED:
+      console.log('Connection dApp not connected. Please call the "connect" function.');
+      break;
+   case RPC_ERROR:
+    console.log('There was an error when broadcasting this transaction to the network.');
+    break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hash is valid
+** Format parameters
+*** If `byteArray`, convert to appropriate hexstring based on assumed format
+* Broadcast the RPC request
+* Return the RPC response results
+
+Example
+<pre>
+{
+  script: '8h89fh398f42f.....89hf2894hf9834',
+  state: 'HALT, BREAK',
+  gas_consumed: '0.13',
+  stack: [
+    {
+      type: 'Integer',
+      value: '1337'
+    }
+  ]
+}
+</pre>
+
+
+====getBlock====
+Get information about a specific block.
+
+
+'''Method Interface'''
+
+<pre>
+function getBlock({
+  blockHeight: integer,
+  network?: string
+}): Promise<{result: BlockDetails}>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface GetBlockInputArgs {
+  blockHeight: integer;
+  network?: string;
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface BlockDetails {
+  hash: string; // Block hash
+  size: number; // Block size (bytes)
+  version: number; // The version number of the block execution
+  previousblockhash: string; // Previous block Hash
+  merkleroot: string; // Merkel root
+  time: number; // Block generation timestamp
+  index: number; // Block index (height)
+  nonce: string; // Block pseudo-random number
+  nextconsensus: string; // Next master biller
+  script: ScriptDetails; // Block call signature authentication information
+  tx: BlockTransactionDetails[]; // Block containing trading group
+  confirmations: number; // Confirmation number (number of blocks after this block)
+  nextblockhash: string; // Next block hash
```

Comment:
I think that, for a greater compatibility with Neo2, Ont, Neo3, and future chains, it's better to reduce required information on blocks here.
Some things are "global" for blockchain tech, like: hash, size, height (index), time, tx array, prevhash, nexthash, confirmations.
Some other things may be different or non-existing: nextconsensus, merkleroot, script, ... perhaps too platform-specific, no?
Perhaps there could be an `extra` field, which is less rigid, and put these things on `extra`, allowing a solid base of information to exist, and protocol-specific stuff on extra.

Reply:
This proposal was created for compatibility with the existing neo platform. These return attributes are directly from the rpc method getblock:
https://docs.neo.org/docs/en-us/reference/rpc/latest-version/api/getblock.html

Once neo3 is out, we may want to look to amend this proposal or make a new one.

Reply:
Only neo2 then.. fine.